test/test_gather.cpp

```diff
@@ -555,4 +584,399 @@ TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+}
+
+// take_along_dim to broadcast, squeeze, then normalization. Segmented
+// as the input dim to take_along_dim cannot be scheduled by the
+// reduction tv
+TEST_F(IndexingOpTest, TakeAlongAxisIntermediateTensorNormalization2_CUDA) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  std::vector<int64_t> shape({32, 1024});
+
+  auto tv0 = makeSymbolicTensor(2);
+  fusion.addInput(tv0);
+  auto tv1 = makeSymbolicTensor(1, DataType::Int);
+  fusion.addInput(tv1);
+
+  auto tv2 = add(tv0, IrBuilder::create<Double>(1));
+  auto tv3 = broadcast(tv1, {false, true});
+  auto tv4 = take_along_axis(tv2, tv3, 1);
+  auto tv5 = squeeze(tv4, std::vector<bool>{false, true});
+  auto tv6 = sum(tv5, {0});
+  auto tv7 = broadcast(tv6, {true});
+  auto tv8 = div(tv5, tv7);
+  fusion.addOutput(tv8);
+
+  at::manual_seed(0);
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  auto options_i = at::TensorOptions().dtype(at::kLong).device(at::kCUDA, 0);
+  auto t0 = at::randn(shape, options);
+  auto t1 = at::randint(0, shape[1], {shape[0]}, options_i);
+  std::vector<c10::IValue> aten_inputs = {t0, t1};
+
+  FusionExecutorCache fec(std::move(fusion_ptr));
+  auto outputs = fec.runFusionWithInputs(aten_inputs);
+
+  validateSegmentation(
+      fec.getMostRecentKernelRuntime(),
+      {ScheduleHeuristic::PointWise, ScheduleHeuristic::Persistent});
+
+  auto t5 = at::take_along_dim(t0.to(at::kDouble) + 1, t1.unsqueeze(-1), 1)
+                .squeeze(1);
+  auto ref = t5 / t5.sum({0}).unsqueeze(0);
+
+  testValidate(&fusion, outputs, aten_inputs, {ref}, __LINE__, __FILE__);
+}
+
+// Normalization, then take_along_axis, then reduction. Similar
+// pattern as cross entropy.
+TEST_F(
+    IndexingOpTest,
+    TakeAlongAxisIntermediateTensorNormalizationAndReduction_CUDA) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  std::vector<int64_t> shape({32, 1024});
+
+  auto tv0 = makeSymbolicTensor(2);
+  fusion.addInput(tv0);
+  auto tv1 = makeSymbolicTensor(2, DataType::Int);
+  fusion.addInput(tv1);
+
+  auto tv2 = sum(tv0, {1});
+  auto tv3 = broadcast(tv2, {false, true});
+  auto tv4 = div(tv0, tv3);
+  auto tv5 = take_along_axis(tv4, tv1, 1);
+  auto tv6 = sum(tv5, {0, 1});
+  fusion.addOutput(tv6);
+
+  at::manual_seed(0);
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  auto options_i = at::TensorOptions().dtype(at::kLong).device(at::kCUDA, 0);
+  auto t0 = at::randn(shape, options);
+  auto t1 = at::randint(0, shape[1], {shape[0], 1}, options_i);
+  std::vector<c10::IValue> aten_inputs = {t0, t1};
+
+  FusionExecutorCache fec(std::move(fusion_ptr));
+  auto outputs = fec.runFusionWithInputs(aten_inputs);
+
+  // The reduction patterns of the normalization and the final
+  // reduction are different, so they are segmented out
```

Comment:
If it is the same pattern, will they be fused? It would be nice to have a test for it.

Reply:
I don't remember. It's actually not specific to `take_along_axis`, but rather it's if the persistent scheduler can fuse the second reduction. I'll add a test for it.

Reply:
It does work. Added `TakeAlongAxisIntermediateTensorNormalizationAndReduction`

Reply:
Great! Very happy to see this type of fusion.